File: 238.product-of-array-except-self.java
// @leet start

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

class Solution {
	public int[] productExceptSelf(int[] nums) {
		List<Integer> left = new ArrayList<>(Collections.nCopies(nums.length, 1));
		List<Integer> right = new ArrayList<>(Collections.nCopies(nums.length, 1));
		for (int i = 1; i < nums.length; i++) {
			left.set(i, (left.get(i - 1) * nums[i - 1]));
		}
		for (int i = nums.length - 2; i >= 0; i--) {
			right.set(i, (right.get(i + 1) * nums[i + 1]));
		}
		int[] answer = new int[nums.length];
		for (int i = 0; i < nums.length; i++) {
			answer[i] = left.get(i) * right.get(i);
		}
		return answer;
	}
}
// @leet end

File: 238.product-of-array-except-self.ts
// @leet start
function productExceptSelf(nums: number[]): number[] {
	const forward = Array(nums.length).fill(1);
	const backward = Array(nums.length).fill(1);
	for (let i = 1; i < nums.length; i++) {
		forward[i] = forward[i - 1] * nums[i-1];
	}
	for (let i = nums.length - 2; i >= 0; i--) {
		backward[i] = backward[i + 1] * nums[i + 1];
	}
	const answer = Array(nums.length);
	for (let i = 0; i < nums.length; i++) {
		answer[i] = forward[i] * backward[i];
	}
	return answer;
}
// @leet end
